File: design_handoff/design/ui_kits/biohub/charts.js
// charts.js — lightweight canvas widgets for the BioHub kit (no deps)
// Mirrors the repo's Vico line charts + adds an EEG band meter.

function risLineChart(canvas, data, opts = {}) {
  const dpr = window.devicePixelRatio || 1;
  const w = canvas.clientWidth, h = canvas.clientHeight;
  canvas.width = w * dpr; canvas.height = h * dpr;
  const ctx = canvas.getContext('2d'); ctx.scale(dpr, dpr);
  const color = opts.color || '#16e0e0';
  const pad = { l: 6, r: 6, t: 10, b: 10 };
  const min = Math.min(...data), max = Math.max(...data), range = (max - min) || 1;
  const xs = i => pad.l + (i / (data.length - 1)) * (w - pad.l - pad.r);
  const ys = v => pad.t + (1 - (v - min) / range) * (h - pad.t - pad.b);
  ctx.clearRect(0, 0, w, h);
  // baseline grid
  ctx.strokeStyle = 'rgba(255,255,255,0.05)'; ctx.lineWidth = 1;
  for (let g = 0; g <= 3; g++) { const y = pad.t + g/3*(h-pad.t-pad.b); ctx.beginPath(); ctx.moveTo(pad.l,y); ctx.lineTo(w-pad.r,y); ctx.stroke(); }
  // area fill
  ctx.beginPath(); ctx.moveTo(xs(0), ys(data[0]));
  data.forEach((v, i) => ctx.lineTo(xs(i), ys(v)));
  ctx.lineTo(xs(data.length-1), h-pad.b); ctx.lineTo(xs(0), h-pad.b); ctx.closePath();
  const grad = ctx.createLinearGradient(0, pad.t, 0, h);
  grad.addColorStop(0, color + '33'); grad.addColorStop(1, color + '00');
  ctx.fillStyle = grad; ctx.fill();
  // line
  ctx.beginPath(); ctx.moveTo(xs(0), ys(data[0]));
  data.forEach((v, i) => ctx.lineTo(xs(i), ys(v)));
  ctx.strokeStyle = color; ctx.lineWidth = 1.5; ctx.lineJoin = 'round'; ctx.stroke();
  // last point dot
  ctx.beginPath(); ctx.arc(xs(data.length-1), ys(data[data.length-1]), 2.5, 0, 7); ctx.fillStyle = color; ctx.fill();
}

// EEG band powers: delta/theta/alpha/beta/gamma vertical bars
function risEegBands(canvas, bands, opts = {}) {
  const dpr = window.devicePixelRatio || 1;
  const w = canvas.clientWidth, h = canvas.clientHeight;
  canvas.width = w * dpr; canvas.height = h * dpr;
  const ctx = canvas.getContext('2d'); ctx.scale(dpr, dpr);
  const labels = ['δ','θ','α','β','γ'];
  const cols = ['#b06bff','#16e0e0','#2fe48a','#f2e205','#ff2d3c'];
  const pad = { l: 4, r: 4, t: 8, b: 16 };
  const bw = (w - pad.l - pad.r) / bands.length;
  const max = Math.max(...bands) || 1;
  ctx.clearRect(0, 0, w, h);
  bands.forEach((v, i) => {
    const bh = (v / max) * (h - pad.t - pad.b);
    const x = pad.l + i * bw + bw * 0.18, bwi = bw * 0.64;
    // track
    ctx.fillStyle = 'rgba(255,255,255,0.04)'; ctx.fillRect(x, pad.t, bwi, h - pad.t - pad.b);
    // bar
    ctx.fillStyle = cols[i]; ctx.fillRect(x, pad.t + (h - pad.t - pad.b - bh), bwi, bh);
    ctx.shadowColor = cols[i]; ctx.shadowBlur = 8; ctx.fillRect(x, pad.t + (h - pad.t - pad.b - bh), bwi, 2); ctx.shadowBlur = 0;
    // label
    ctx.fillStyle = '#98a3a5'; ctx.font = '11px "JetBrains Mono", monospace'; ctx.textAlign = 'center';
    ctx.fillText(labels[i], x + bwi/2, h - 4);
  });
}

window.risLineChart = risLineChart; window.risEegBands = risEegBands;

// 4-channel live EEG waveform (TP9/AF7/AF8/TP10) — mirrors MuseRecorder EegWaveformCard
function risEegWaveform(canvas, channels, opts = {}) {
  const dpr = window.devicePixelRatio || 1;
  const w = canvas.clientWidth, h = canvas.clientHeight;
  canvas.width = w * dpr; canvas.height = h * dpr;
  const ctx = canvas.getContext('2d'); ctx.scale(dpr, dpr);
  const cols = ['#2fe48a', '#16e0e0', '#e08a3c', '#b06bff']; // TP9 AF7 AF8 TP10
  ctx.clearRect(0, 0, w, h);
  const chH = h / channels.length;
  channels.forEach((samples, idx) => {
    if (!samples || samples.length < 2) return;
    const yBase = idx * chH;
    const min = Math.min(...samples), max = Math.max(...samples), range = Math.max(max - min, 20);
    const xStep = w / (samples.length - 1);
    if (idx > 0) { ctx.strokeStyle = cols[idx] + '26'; ctx.lineWidth = 0.5; ctx.beginPath(); ctx.moveTo(0, yBase); ctx.lineTo(w, yBase); ctx.stroke(); }
    ctx.beginPath();
    samples.forEach((v, i) => {
      const x = i * xStep, y = yBase + chH * (1 - Math.min(Math.max((v - min) / range, 0), 1)) * 0.9 - chH * 0.05;
      i === 0 ? ctx.moveTo(x, y) : ctx.lineTo(x, y);
    });
    ctx.strokeStyle = cols[idx]; ctx.lineWidth = 1.3; ctx.lineJoin = 'round'; ctx.stroke();
  });
}

// synth EEG channel data (for the prototype's live feed)
function risSynthEeg(n, seed, amp) {
  const a = []; for (let i = 0; i < n; i++) a.push(Math.sin(i * 0.4 + seed) * amp + Math.sin(i * 1.3 + seed * 2) * amp * 0.4 + (Math.random() - 0.5) * amp * 0.5);
  return a;
}

window.risEegWaveform = risEegWaveform; window.risSynthEeg = risSynthEeg;

File: design_handoff/design/compose/Charts.kt
package com.relic.ris.theme

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.dp
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.sin

/**
 * RELIC INTERFACE SYSTEM — Canvas chart composables.
 * Kotlin port of charts.js (risLineChart / risEegBands) for BioHub.
 * The repo uses Vico; these are dependency-free RIS-styled equivalents for the
 * cases Vico is overkill, or as a reference for theming Vico to RIS.
 */

/** Line + area chart over a series of values. */
@Composable
fun RisLineChart(
    data: List<Float>,
    modifier: Modifier = Modifier,
    color: androidx.compose.ui.graphics.Color = RisCyan,
) {
    if (data.size < 2) return
    Canvas(modifier) {
        val padX = 6f; val padY = 10f
        val w = size.width; val h = size.height
        val min = data.min(); val max = data.max(); val range = (max - min).takeIf { it != 0f } ?: 1f
        fun x(i: Int) = padX + i.toFloat() / (data.size - 1) * (w - padX * 2)
        fun y(v: Float) = padY + (1f - (v - min) / range) * (h - padY * 2)
        // baseline grid
        for (g in 0..3) {
            val gy = padY + g / 3f * (h - padY * 2)
            drawLine(RisLineFaint, Offset(padX, gy), Offset(w - padX, gy), 1f)
        }
        // area
        val area = Path().apply {
            moveTo(x(0), y(data[0])); data.forEachIndexed { i, v -> lineTo(x(i), y(v)) }
            lineTo(x(data.size - 1), h - padY); lineTo(x(0), h - padY); close()
        }
        drawPath(area, color.copy(alpha = 0.20f))
        // line
        val line = Path().apply { moveTo(x(0), y(data[0])); data.forEachIndexed { i, v -> lineTo(x(i), y(v)) } }
        drawPath(line, color, style = Stroke(width = 3f))
        drawCircle(color, radius = 4f, center = Offset(x(data.size - 1), y(data.last())))
    }
}

/** EEG band-power bars: delta/theta/alpha/beta/gamma. Expects 5 normalized values. */
@Composable
fun RisEegBands(bands: List<Float>, modifier: Modifier = Modifier) {
    val cols = listOf(RisViolet, RisCyan, RisGreen, RisAmber, RisRed)
    Canvas(modifier) {
        val padX = 4f; val padTop = 8f; val padBottom = 16f
        val n = bands.size.coerceAtLeast(1)
        val bw = (size.width - padX * 2) / n
        val max = (bands.maxOrNull() ?: 1f).takeIf { it > 0f } ?: 1f
        bands.forEachIndexed { i, v ->
            val bh = v / max * (size.height - padTop - padBottom)
            val x = padX + i * bw + bw * 0.18f
            val barW = bw * 0.64f
            // track
            drawRect(RisLineFaint, topLeft = Offset(x, padTop), size = androidx.compose.ui.geometry.Size(barW, size.height - padTop - padBottom))
            // bar
            drawRect(cols[i % cols.size], topLeft = Offset(x, padTop + (size.height - padTop - padBottom - bh)), size = androidx.compose.ui.geometry.Size(barW, bh))
        }
    }
}

/** Live multi-channel EEG waveform (e.g. Muse TP9/AF7/AF8/TP10). */
@Composable
fun RisEegWaveform(
    channels: List<List<Float>>,
    modifier: Modifier = Modifier,
    colors: List<androidx.compose.ui.graphics.Color> = listOf(RisCyan, RisAmber, RisGreen, RisViolet),
) {
    Canvas(modifier) {
        val rows = channels.size.coerceAtLeast(1)
        val rowH = size.height / rows
        channels.forEachIndexed { ci, series ->
            if (series.size < 2) return@forEachIndexed
            val baseY = rowH * ci + rowH / 2
            val amp = rowH * 0.4f
            val path = Path()
            series.forEachIndexed { i, v ->
                val px = i.toFloat() / (series.size - 1) * size.width
                val py = baseY - v.coerceIn(-1f, 1f) * amp
                if (i == 0) path.moveTo(px, py) else path.lineTo(px, py)
            }
            drawPath(path, colors[ci % colors.size], style = Stroke(width = 1.5f))
        }
    }
}

/**
 * Decorative voice-activity indicator — animated symmetric bars (NOT real
 * amplitude). Idle = faint flat line; [active] drives the motion. RIS HUD style:
 * one accent [tone], hard round-capped bars. Size via [modifier] (e.g. height 46.dp).
 */
@Composable
fun RisVoiceWave(
    active: Boolean,
    modifier: Modifier = Modifier,
    tone: androidx.compose.ui.graphics.Color = RisCyan,
    bars: Int = 26,
) {
    val phase by rememberInfiniteTransition(label = "ris-voice").animateFloat(
        initialValue = 0f,
        targetValue = 2f * PI.toFloat(),
        animationSpec = infiniteRepeatable(tween(1100, easing = LinearEasing), RepeatMode.Restart),
        label = "ris-voice-phase",
    )
    Canvas(modifier) {
        val mid = size.height / 2f
        val step = size.width / bars
        val sw = step * 0.45f
        for (i in 0 until bars) {
            val x = i * step + step / 2f
            val a = if (active) 0.18f + 0.82f * abs(sin(phase + i * 0.55f)) else 0.10f
            val bh = mid * a
            drawLine(tone, Offset(x, mid - bh), Offset(x, mid + bh), strokeWidth = sw, cap = StrokeCap.Round)
        }
    }
}
